packages/rest/src/body-parser.ts

```diff
@@ -0,0 +1,287 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {OperationObject} from '@loopback/openapi-v3-types';
+import * as debugModule from 'debug';
+import * as HttpErrors from 'http-errors';
+import {Request, Response, RequestBodyParserOptions} from './types';
+
+import {
+  json,
+  urlencoded,
+  text,
+  OptionsJson,
+  OptionsUrlencoded,
+  OptionsText,
+  Options,
+} from 'body-parser';
+import {inject} from '@loopback/context';
+import {SchemaObject, ReferenceObject, isReferenceObject} from '..';
+import {RestBindings} from './keys';
+import {is} from 'type-is';
+import {RestHttpErrors} from './rest-http-error';
+
+type HttpError = HttpErrors.HttpError;
+
+const debug = debugModule('loopback:rest:body-parser');
+
+export type RequestBody = {
+  // tslint:disable:no-any
+  value: any | undefined;
+  coercionRequired?: boolean;
+  mediaType?: string;
+  schema?: SchemaObject | ReferenceObject;
+};
+
+/**
+ * Get the content-type header value from the request
+ * @param req Http request
+ */
+function getContentType(req: Request): string | undefined {
+  return req.get('content-type');
+}
+
+/**
+ * Express body parser function type
+ */
+type BodyParserWithCallback = (
+  request: Request,
+  response: Response,
+  callback: (err: HttpError) => void,
+) => void;
+
+function normalizeParsingError(err: HttpError) {
+  debug('Cannot parse request body %j', err);
+  if (!err.statusCode || err.statusCode >= 500) {
+    err.statusCode = 400;
+  }
+  return err;
+}
+
+/**
+ * Parse the body asynchronously
+ * @param handle The express middleware handler
+ * @param request Http request
+ */
+function parse(
+  handle: BodyParserWithCallback,
+  request: Request,
+): Promise<void> {
```

Comment:
Can we change `parse` to return the parsed body please?

I find such design more cohesive: the details about request parsing and storing the parsed body on `request.body` property stay inside `parse`. The rest of this file can stay free from this (not particularly nice) detail.